runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/processing/failures/WorkFailureProcessor.java

```diff
@@ -103,37 +126,38 @@ private void executeWithDelay(long delayMs, Work work) {
   }
 
   private boolean shouldRetryLocally(String computationId, Work work, Throwable t) {
-    if (KeyTokenInvalidException.isKeyTokenInvalidException(t)) {
+    Throwable parsedException = t instanceof UserCodeException ? t.getCause() : t;
+    if (KeyTokenInvalidException.isKeyTokenInvalidException(parsedException)) {
       LOG.debug(
           "Execution of work for computation '{}' on key '{}' failed due to token expiration. "
               + "Work will not be retried locally.",
           computationId,
           work.getWorkItem().getKey().toStringUtf8());
-    } else if (WorkItemCancelledException.isWorkItemCancelledException(t)) {
+    } else if (WorkItemCancelledException.isWorkItemCancelledException(parsedException)) {
       LOG.debug(
           "Execution of work for computation '{}' on key '{}' failed. "
               + "Work will not be retried locally.",
           computationId,
           work.getWorkItem().getShardingKey());
     } else {
-      LastExceptionDataProvider.reportException(t);
+      LastExceptionDataProvider.reportException(parsedException);
       LOG.debug("Failed work: {}", work);
       Duration elapsedTimeSinceStart = new Duration(work.getStartTime(), clock.get());
-      if (!failureReporter.reportFailure(computationId, work.getWorkItem(), t)) {
+      if (!failureTracker.trackFailure(computationId, work.getWorkItem(), parsedException)) {
         LOG.error(
             "Execution of work for computation '{}' on key '{}' failed with uncaught exception, "
                 + "and Windmill indicated not to retry locally.",
             computationId,
             work.getWorkItem().getKey().toStringUtf8(),
-            t);
-      } else if (isOutOfMemoryError(t)) {
+            parsedException);
+      } else if (isOutOfMemoryError(parsedException)) {
         LOG.error(
             "Execution of work for computation '{}' for key '{}' failed with out-of-memory. "
                 + "Work will not be retried locally. Heap dump {}.",
             computationId,
             work.getWorkItem().getKey().toStringUtf8(),
-            getHeapDumpLog(),
-            t);
+            dumpHeap(),
```

Comment:
Can we extract this to a separate line to highlight that it is dumping heap?

Reply:
done